File: patterns/stack/pseudo_stack/is_valid.cpp
/*
Given a list of same symbols, we need to check if every opening symbol
has its closing symbol.

e.g.
symbols = ["(",")","(","(",")",")"]
output = True

symbols = ["(","(",")"]
output = False

This approach really does not need the stack, we can just use a counter kek.
*/
#include <iostream>
#include <vector>

using namespace std;

bool is_valid(const vector<char> &symbols)
{
    int counter = 0;

    for (char symbol : symbols)
    {
        if (symbol == '(')
            counter++;
        else
            counter--;

        // Fast exit
        if (counter < 0)
            return false;
    }

    return counter == 0;
}

int main()
{
    vector<char> symbols = {'(', ')', '(', '(', ')', ')'};
    vector<char> bad_symbols = {'(', '(', ')'};

    cout << "Symbols: " << endl;
    for (char symbol : symbols)
    {
        cout << symbol << " ";
    }

    cout << "\nIs valid: " << (is_valid(symbols) ? "True" : "False") << endl;

    cout << "\nBad symbols: " << endl;
    for (char symbol : bad_symbols)
    {
        cout << symbol << " ";
    }

    cout << "\nIs valid: " << (is_valid(bad_symbols) ? "True" : "False") << endl;
}

File: patterns/stack/pseudo_stack/is_valid.py
"""
Given a list of same symbols, we need to check if every opening symbol
has its closing symbol.

e.g.
symbols = ["(",")","(","(",")",")"]
output = True

symbols = ["(","(",")"]
output = False

This approach really does not need the stack, we can just use a counter kek.
"""


def is_valid(symbols: list[str]) -> bool:
    counter: int = 0
    for symbol in symbols:
        if symbol == "(":
            counter += 1
        else:
            counter -= 1

        if counter < 0:
            return False

    return counter == 0


if __name__ == "__main__":
    symbols: list[str] = ["(", ")", "(", "(", ")", ")"]
    bad_symbols: list[str] = ["(", "(", ")"]

    print(f"Symbols: {symbols} -> {is_valid(symbols)}")
    print(f"Bad symbols: {bad_symbols} -> {is_valid(bad_symbols)}")
